Mathlib/Probability/Kernel/RadonNikodym.lean

```diff
@@ -0,0 +1,443 @@
+          + withDensity (κ + η) (fun a x ↦ Real.toNNReal (rnDerivAux κ (κ + η) a x))) a s
+        = κ a s + η a s := by
+      rw [this]
+      simp
+    simp only [coeFn_add, Pi.add_apply, Measure.add_toOuterMeasure, OuterMeasure.coe_add]
+      at h
+    rwa [withDensity_rnDerivAux, add_comm, ENNReal.add_right_inj (measure_ne_top _ _)] at h
+  have : ∀ b, (Real.toNNReal b : ℝ≥0∞) = ENNReal.ofReal b := fun _ ↦ rfl
+  simp_rw [this, ENNReal.ofReal_sub _ (rnDerivAux_nonneg h_le), ENNReal.ofReal_one]
+  rw [withDensity_sub_add_cancel]
+  · rw [withDensity_one']
+  · exact measurable_const
+  · exact (measurable_rnDerivAux _ _).ennreal_ofReal
+  · refine fun a ↦ ae_of_all _ (fun x ↦ ?_)
+    simp only [ENNReal.ofReal_le_one]
+    exact density_le_one ((fst_map_id_prod _ measurable_const).trans_le h_le) _ _ _
+
+/-- A set of points in `α × γ` related to the absolute continuity / mutual singularity of
+`κ` and `η`. -/
+def mutuallySingularSet (κ η : kernel α γ) : Set (α × γ) := {p | rnDerivAux κ (κ + η) p.1 p.2 = 1}
+
+/-- A set of points in `α × γ` related to the absolute continuity / mutual singularity of
+`κ` and `η`. That is,
+* `withDensity η (rnDeriv κ η) a (mutuallySingularSetSlice κ η a) = 0`,
+* `singularPart κ η a (mutuallySingularSetSlice κ η a)ᶜ = 0`.
+ -/
+def mutuallySingularSetSlice (κ η : kernel α γ) (a : α) : Set γ :=
+  {x | rnDerivAux κ (κ + η) a x = 1}
+
+lemma measurableSet_mutuallySingularSet (κ η : kernel α γ) :
+    MeasurableSet (mutuallySingularSet κ η) :=
+  measurable_rnDerivAux κ (κ + η) (measurableSet_singleton 1)
+
+lemma measurableSet_mutuallySingularSetSlice (κ η : kernel α γ) (a : α) :
+    MeasurableSet (mutuallySingularSetSlice κ η a) :=
+  measurable_prod_mk_left (measurableSet_mutuallySingularSet κ η)
+
+lemma measure_mutuallySingularSetSlice (κ η : kernel α γ) [IsFiniteKernel κ] [IsFiniteKernel η]
+    (a : α) :
+    η a (mutuallySingularSetSlice κ η a) = 0 := by
+  have h_coe : ∀ b, (Real.toNNReal b : ℝ≥0∞) = ENNReal.ofReal b := fun _ ↦ rfl
+  suffices withDensity (κ + η) (fun a x ↦ Real.toNNReal
+      (1 - rnDerivAux κ (κ + η) a x)) a {x | rnDerivAux κ (κ + η) a x = 1} = 0 by
+    rwa [withDensity_one_sub_rnDerivAux κ η] at this
+  simp_rw [h_coe]
+  rw [kernel.withDensity_apply', lintegral_eq_zero_iff, EventuallyEq, ae_restrict_iff]
+  rotate_left
+  · exact (measurable_const.sub
+      ((measurable_rnDerivAux _ _).comp measurable_prod_mk_left)).ennreal_ofReal
+      (measurableSet_singleton _)
+  · exact (measurable_const.sub
+      ((measurable_rnDerivAux _ _).comp measurable_prod_mk_left)).ennreal_ofReal
+  · exact (measurable_const.sub (measurable_rnDerivAux _ _)).ennreal_ofReal
+  refine ae_of_all _ (fun x hx ↦ ?_)
+  simp only [mem_setOf_eq] at hx
+  simp [hx]
+
+/-- Radon-Nikodym derivative of the kernel `κ` with respect to the kernel `η`. -/
+noncomputable
+def rnDeriv (κ η : kernel α γ) (a : α) (x : γ) : ℝ≥0∞ :=
```

Comment:
maybe
```suggestion
irreducible_def rnDeriv (κ η : kernel α γ) (a : α) (x : γ) : ℝ≥0∞ :=
```
Just to hide forever the details of the messy construction from the kernel. It will force you to unfold by hand the definition in the first few proofs below, but I guess that once the API is done you will never use the bare definition again.